packages/explorer-helpers/src/readMultipartWebStream.ts

```diff
@@ -0,0 +1,174 @@
+import { Observable } from 'zen-observable-ts';
+import type { GraphQLError } from 'graphql';
+import type MIMEType from 'whatwg-mimetype';
+import type { ResponseData } from './postMessageRelayHelpers';
+
+export interface MultipartResponse {
+  data: ResponseData & {
+    incremental?: Array<
+      ResponseData & { path: NonNullable<ResponseData['path']> }
+    >;
+    error?: { message: string; stack?: string };
+    hasNext?: boolean;
+  };
+  headers?: Record<string, string> | Record<string, string>[];
+  size: number;
+}
+
+// https://apollographql.quip.com/mkWRAJfuxa7L/Multipart-subscriptions-protocol-spec
+export interface MultipartSubscriptionResponse {
+  data: {
+    errors?: Array<GraphQLError>;
+    payload:
+      | (ResponseData & {
+          error?: { message: string; stack?: string };
+        })
+      | null;
+  };
+  headers?: Record<string, string> | Record<string, string>[];
+  size: number;
+  // True if --graphql-- message boundary is in the response
+  shouldTerminate?: boolean;
+}
+
+export function readMultipartWebStream(response: Response, mimeType: MIMEType) {
+  if (response.body === null) {
+    throw new Error('Missing body');
+  } else if (typeof response.body.tee !== 'function') {
+    // not sure if we actually need this check in explorer?
+    throw new Error(
+      'Streaming bodies not supported by provided fetch implementation'
+    );
+  }
+
+  const decoder = new TextDecoder('utf-8');
+  let buffer = '';
+
+  const messageBoundary = `--${mimeType.parameters.get('boundary') || '-'}`;
+  const subscriptionTerminationMessageBoundary = '--graphql--';
+
+  const reader = response.body.getReader();
+  return {
+    closeReadableStream: () => reader.cancel(),
+    observable: new Observable<
+      MultipartResponse | MultipartSubscriptionResponse
```

Comment:
can we add in the `closeReadableStream` stuff? 
```suggestion
      (MultipartResponse | MultipartSubscriptionResponse)  & {
      closeReadableStream: () => void;
    }
```

Reply:
Ah these are slightly different in Explorer & the embeds.

`closeReadableStream` is exposed in the return object in this version which I think is cleaner, since its the same for every response

Reply:
Ooohh I totally missed it above! agreed! I like this! :)